state/protocol/badger/mutator.go

```diff
@@ -576,6 +589,20 @@ func (m *FollowerState) insert(ctx context.Context, candidate *flow.Block, certi
 			}
 		}
 
+		if hasChanges {
+			err = m.protocolState.StoreTx(updatedStateID, updatedState)(tx)
+			// in case of fork, the protocol state may already exist
+			if err != nil && !errors.Is(err, storage.ErrAlreadyExists) {
+				return fmt.Errorf("could not store protocol state (%v): %w", updatedStateID, err)
+			}
+		}
+
+		err = m.protocolState.Index(blockID, updatedStateID)(tx)
+		if err != nil {
+			return fmt.Errorf("could not index protocol state (%v) for block (%v): %w",
+				updatedStateID, blockID, err)
+		}
+
```

Comment:
In my opinion, the bulk of this code should go _before_ line https://github.com/onflow/flow-go/blob/79b2ebb22555484a72a659b5d8677d0a73715179/state/protocol/badger/mutator.go#L539

Reasoning:

* Both [`ProtocolState.StoreTx`](https://github.com/onflow/flow-go/blob/79b2ebb22555484a72a659b5d8677d0a73715179/state/protocol/badger/mutator.go#L593) as well as [`ProtocolState.Index`](https://github.com/onflow/flow-go/blob/79b2ebb22555484a72a659b5d8677d0a73715179/state/protocol/badger/mutator.go#L600) return anonymous functions for _later_ execution as part of a badger transaction.  In particular, they might front-load a significant amount of computation and sanity checks as pre-processing outside of their transaction fragment. 

In my mind, the code [lines 526-607](https://github.com/onflow/flow-go/blob/79b2ebb22555484a72a659b5d8677d0a73715179/state/protocol/badger/mutator.go#L526-L607) could look something like this:
```golang
updatedState, updatedStateID, hasChanges := protocolStateUpdater.Build()
// TODO: check if updatedStateID corresponds to the root protocol state ID stored in payload
// if updatedStateID != payload.ProtocolStateID {
// 	return state.NewInvalidExtension("invalid protocol state transition detected expected (%x) got %x", payload.ProtocolStateID, updatedStateID)
// }
if hasChanges {
  dbUpdates = append(dbUpdates, m.protocolState.StoreTx(updatedStateID, updatedState))
}
dbUpdates = append(dbUpdates, m.protocolState.Index(blockID, updatedStateID))

// events is a queue of node-internal events (aka notifications) that are emitted after the database write succeeded
var events []func()

if certifyingQC != nil {
  dbUpdates = append(dbUpdates, m.qcs.StoreTx(certifyingQC))

  // queue an BlockProcessable event for candidate block, since it is certified
  events = append(events, func() {
    m.consumer.BlockProcessable(candidate.Header, certifyingQC)
  })
}

// Both the header itself and its payload are in compliance with the protocol state.
// We can now store the candidate block, as well as adding its final seal
// to the seal index and initializing its children index.
qc := candidate.Header.QuorumCertificate()
err = operation.RetryOnConflictTx(m.db, transaction.Update, func(tx *transaction.Tx) error {
  // insert the block into the database AND cache
  err := m.blocks.StoreTx(candidate)(tx)
  if err != nil {
    return fmt.Errorf("could not store candidate block: %w", err)
  }

  err = m.qcs.StoreTx(qc)(tx)
  if err != nil {
    if !errors.Is(err, storage.ErrAlreadyExists) {
      return fmt.Errorf("could not store incorporated qc: %w", err)
    }
  } else {
    // trigger BlockProcessable for parent blocks above root height
    if parent.Height > m.finalizedRootHeight {
      events = append(events, func() {
        m.consumer.BlockProcessable(parent, qc)
      })
    }
  }

  // index the latest sealed block in this fork
  err = transaction.WithTx(operation.IndexLatestSealAtBlock(blockID, latestSealID))(tx)
  if err != nil {
    return fmt.Errorf("could not index candidate seal: %w", err)
  }

  // index the child block for recovery
  err = transaction.WithTx(procedure.IndexNewBlock(blockID, candidate.Header.ParentID))(tx)
  if err != nil {
    return fmt.Errorf("could not index new block: %w", err)
  }

  // apply any optional DB operations from service events
  for _, apply := range dbUpdates {
    err := apply(tx)
    if err != nil {
      return fmt.Errorf("could not apply operation: %w", err)
    }
  }
  return nil
})
```

_Note:_
similar argument could apply to other operations in the [body of `operation.RetryOnConflictTx`](https://github.com/onflow/flow-go/blob/79b2ebb22555484a72a659b5d8677d0a73715179/state/protocol/badger/mutator.go#L539)

Reply:
[`fb23b0d` (#4613)](https://github.com/onflow/flow-go/pull/4613/commits/fb23b0d37817a9b7f04212efeeefd427b44fed6c)